### DISCUSSION TITLE
How to check for deprecations with PHPUnit tests

### DISCUSSION THREAD

**Discussion by jnoordsij** [laravel/framework#45531] [Q&A] [answered]
To prepare for the new PHP 8.2 release, I was running my applications and test suites on the latest RC. Using the `LOG_DEPRECATIONS_CHANNEL` you can easily log any deprecations to your favorite channel when running your application.

However when running the test suite, I am unable to check wether or not any deprecations were triggered during the tests. It seems that the test app container seems to catch all deprecations and handle them internally, so there's no way to find any trace of them.

Anyone has any suggestions on how to make the deprecations show up in the PHPUnit output (or any other place even?) when running tests?

**Comment by nunomaduro** [comment, marked as answer]
Check this: https://laravel.com/docs/9.x/http-tests#exception-handling.

**Reply by jnoordsij** [reply]
Ahh thanks, must've missed that because I didn't expect to find this within the `HTTP tests` section specifically. Altough the docs suggest to apply this on an individual per-request base, it looks like you can simply toggle this somewhere within your base `TestCase` class to apply it for all tests.

**Comment by loma** [comment]
I have the same problem, was migrating to laravel 8, i assume laravel 9 should be similar
1. set the APP_ENV to development (the log will skip if the test is running under **testing** env)
2. set the deprecations log channel to **stderr** (https://laravel.com/docs/8.x/logging#logging-deprecation-warnings)

**Comment by mfn** [comment]
phpunit has built-in support for reporting PHP deprecations, this can be controlled via things like `failOnDeprecation`, `restrictDeprecations` (or now renamed to `ignoreIndirectDeprecations`) and `displayDetailsOnTestsThatTriggerDeprecations` (maybe I forgot a setting).

However, all of this doesn't seem to apply because Laravel takes matters in its own hand. Sure you can run it like this:
`LOG_DEPRECATIONS_CHANNEL=errorlog  vendor/bin/phpunit`

**But**: you will only get the log output intertwined with phpunit output that way, but phpunit is not able to "know" about tests being affected and can't mark them as `D` (as in `D`eprecated).

Did anyone figure out how to better have this integrated with PHPUnit?

**Comment by bretto36** [comment]
So i just found this method

```
protected function shouldIgnoreDeprecationErrors()
    {
        return ! class_exists(LogManager::class)
            || ! static::$app->hasBeenBootstrapped()
            || (static::$app->runningUnitTests() && ! Env::get('LOG_DEPRECATIONS_WHILE_TESTING'));
    }
```
    
    i added `LOG_DEPRECATIONS_WHILE_TESTING=true` to me .env.testing  file and it started working again. 
    
LOG_DEPRECATIONS_CHANNEL=stderr
LOG_DEPRECATIONS_TRACE=false
LOG_DEPRECATIONS_WHILE_TESTING=true
